### PR TITLE
Avoid raw pointers in favor of VectorMemory::Pointer in SolverRichardson.

### PR DESCRIPTION
**Pull request by bangerth**
While there, also clean up a couple of other things.

Passes the testsuite. In reference to #4935.

### CODE REVIEW

File: include/deal.II/lac/solver_richardson.h [hunk 1]
```diff
@@ -143,33 +142,17 @@ class SolverRichardson : public Solver<VectorType>
 protected:
   /**
    * Implementation of the computation of the norm of the residual.
+   * Depending on the flags given to the solver, the default
+   * implementation of this function uses either the actual
+   * residual, @p r, or the preconditioner residual, @p d.
```
Comment (Rombur) [inline include/deal.II/lac/solver_richardson.h]: I think there is a typo `preconditioner residual` -> `preconditioned residual`

File: include/deal.II/lac/solver_richardson.h [hunk 2]
```diff
@@ -143,33 +142,17 @@ class SolverRichardson : public Solver<VectorType>
...
-   */
-  VectorType *Vr;
-  /**
-   * Preconditioned residual. Temporary vector allocated through the
-   * VectorMemory object at the start of the actual solution process and
-   * deallocated at the end.
-   */
-  VectorType *Vd;
+  virtual typename VectorType::value_type criterion(const VectorType &r,
+                                                    const VectorType &d) const;
```
Comment (Rombur) [inline include/deal.II/lac/solver_richardson.h]: Do we need an entry in the changelog for this incompatibility? The function is protected...

**Comment by bangerth** [comment]
I fixed the typo and added a changelog entry.

**Comment by masterleinad** [comment]
The failing test seems to be unrelated. It's probably just due to overloading the tester.

**Comment by bangerth** [comment]
I agree. The test did not fail on my system when I ran the testsuite for this patch.

**Comment by masterleinad** [comment]
I'll take that risk. :wink:

**Comment by bangerth** [comment]
Real men take risks!

**Pull request merged**